# Predicting Yearly Salary Model!
**Creating a Machine Learning (ML) model capable of predicting the salaries of workers in programming and data science** can be useful for several reasons:

1. **Informed Decision-Making**: It helps professionals and employers make more informed decisions about compensation. Workers can negotiate fair salaries, and companies can offer competitive packages.

2. **Market Evaluation**: It allows companies to assess their position in the job market. Knowing how their employees' salaries compare to competitors is crucial for retaining talent and attracting new hires.

3. **Resource Optimization**: Companies can use these models to allocate resources more efficiently. For example, they can determine how much to pay a new employee based on their skills and experience.

4. **Identifying Salary Disparities**: It helps identify potential unfair salary gaps. If a model suggests disparities in compensation among employees with similar skills, measures can be taken to address this issue.

5. **Career Planning**: For professionals, having an idea of how their salary might evolve over time can assist in career planning. They can set goals and strategies to improve their income.

6. **Valuable Data Collection**: Building and training these models gathers valuable data about the job market, which can be useful for further studies and analysis.

7. **Transparency and Equity**: It promotes transparency and equity in compensation, as it is based on objective data rather than arbitrary decisions.

8. **Support for Strategic Decision-Making**: Companies can use these models to make strategic decisions about expanding their programming and data science teams, diversifying skills, and long-term planning.

In summary, creating an ML model to predict salaries in programming and data science can provide benefits to both employees and employers, improving fairness, efficiency, and decision-making in the job market.


In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_rows", 50)
pd.set_option("display.max_columns", 50)
pd.options.display.float_format = "{:,.2f}".format

# Import the df created in the '../Exploratory_Analysis/1_Data_Preparation.ipynb'
df = pd.read_pickle("../../data/interim/interest_SO_survey_2023.pkl")


In [2]:
df.head()

Age                                         BuyNewTool  \
0  18-24 years old                                                NaN   
1  25-34 years old  Start a free trial;Ask developers I know/work ...   
2  45-54 years old  Start a free trial;Ask developers I know/work ...   
3  25-34 years old  Start a free trial;Ask developers I know/work ...   
4  25-34 years old  Start a free trial;Ask developers I know/work ...   

                    Country           CoursesCert  \
0                       NaN                   NaN   
1  United States of America                 Other   
2  United States of America                   NaN   
3  United States of America                   NaN   
4               Philippines  Other;Codecademy;edX   

                             CurrentJob  \
0                                   NaN   
1  Senior Executive (C-Suite, VP, etc.)   
2                   Developer, back-end   
3                  Developer, front-end   
4                 Developer, full-stack   

                                        EdLevel  \
0                                           NaN   
1  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                          Employment  \
0                                                NaN   
1                                Employed, full-time   
2                                Employed, full-time   
3                                Employed, full-time   
4  Employed, full-time;Independent contractor, fr...   

                                            Industry  \
0                                                NaN   
1  Information Services, IT, Software Development...   
2  Information Services, IT, Software Development...   
3                                                NaN   
4                                              Other   

                       MainBranch  \
0                   None of these   
1  I am a developer by profession   
2  I am a developer by profession   
3  I am a developer by profession   
4  I am a developer by profession   

                                          OSPersonal  \
0                                                NaN   
1  iOS;iPadOS;MacOS;Windows;Windows Subsystem for...   
2                            MacOS;Other Linux-based   
3                                   iOS;iPadOS;MacOS   
4                            Other (Please Specify):   

                                    OSProffesional  \
0                                              NaN   
1  MacOS;Windows;Windows Subsystem for Linux (WSL)   
2                          MacOS;Other Linux-based   
3                                 iOS;iPadOS;MacOS   
4                          Other (Please Specify):   

                                    ProfessionalTech  \
0                                                NaN   
1  DevOps function;Microservices;Automated testin...   
2  DevOps function;Microservices;Automated testin...   
3  Automated testing;Continuous integration (CI) ...   
4  Microservices;Automated testing;Observability ...   

                             RemoteWork TargetAIDeveloper    TargetAISearch  \
0                                   NaN               NaN               NaN   
1                                Remote    GitHub Copilot  ChatGPT;Neeva AI   
2  Hybrid (some remote, some in-person)               NaN               NaN   
3  Hybrid (some remote, some in-person)               NaN               NaN   
4                                Remote               NaN           ChatGPT   

            TargetCollabTools  \
0                         NaN   
1      Vim;Visual Studio Code   
2                 Emacs;Helix   
3  IntelliJ IDEA;Vim;WebStorm   
4      Vim;Visual Studio Code   

                                    TargetDatabase  \
0                                              NaN   
1            

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87981 entries, 0 to 87980
Data columns (total 39 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                87981 non-null  object 
 1   BuyNewTool         83009 non-null  object 
 2   Country            87973 non-null  object 
 3   CoursesCert        37076 non-null  object 
 4   CurrentJob         76872 non-null  object 
 5   EdLevel            87973 non-null  object 
 6   Employment         87898 non-null  object 
 7   Industry           36774 non-null  object 
 8   MainBranch         87981 non-null  object 
 9   OSPersonal         86557 non-null  object 
 10  OSProffesional     78587 non-null  object 
 11  ProfessionalTech   41783 non-null  object 
 12  RemoteWork         73810 non-null  object 
 13  TargetAIDeveloper  19587 non-null  object 
 14  TargetAISearch     46150 non-null  object 
 15  TargetCollabTools  76649 non-null  object 
 16  TargetDatabase     609

In [4]:
df2 = df.dropna(subset=["YearlySalary"]).copy()
df2.isna().sum()

Age                      0
BuyNewTool            1817
Country                  0
CoursesCert          27275
CurrentJob             115
EdLevel                  0
Employment              12
Industry             20272
MainBranch               0
OSPersonal             485
OSProffesional        2989
ProfessionalTech     16293
RemoteWork              79
TargetAIDeveloper    37357
TargetAISearch       23813
TargetCollabTools     5405
TargetDatabase       12947
TargetFramework      15573
TargetLanguage        3448
TargetLibraries      21678
TargetPlatform       17610
TargetTools           9153
WorkedAIDeveloper    34079
WorkedAISearch       18534
WorkedCollabTools      584
WorkedDatabase        6255
WorkedFramework      10154
WorkedLanguage         136
WorkedLibraries      16250
WorkedPlatform       10644
WorkedTools           3964
WorkingYears         15381
YearlySalary             0
YearsCode               69
YearsCodePro           194
z_BenefitsAI         27414
z_FavorableAI        14901
z

In [5]:
# I prefer to fill this with 'Not_Responded' instead of dropping the column or the Nan values
df2["Industry"] = df2["Industry"].fillna("Not_Responded")
df2["Industry"] = df2["Industry"].apply(
    lambda x: "Technology"
    if x == "Information Services, IT, Software Development, or other Technology"
    else x
)
df2["Industry"] = df2["Industry"].apply(
    lambda x: "Manufacturing"
    if x == "Manufacturing, Transportation, or Supply Chain"
    else x
)
df2["Industry"] = df2["Industry"].apply(
    lambda x: "Retail" if x == "Retail and Consumer Services" else x
)
top_industry = df2["Industry"].value_counts().nlargest(8).index
df2["Industry"] = df2["Industry"].apply(lambda x: x if x in top_industry else "Other")
df2["Industry"].value_counts()

Industry
Not_Responded         20272
Technology            13469
Other                  4715
Financial Services     3441
Manufacturing          2001
Healthcare             1702
Retail                 1552
Higher Education        867
Name: count, dtype: int64

I will drop some features that will no be useful for my model like the target things,or those ones with too many missing values.

In [6]:
columns_to_drop = [
    "BuyNewTool",
    "CoursesCert",
    "ProfessionalTech",
    "TargetAIDeveloper",
    "TargetAISearch",
    "TargetCollabTools",
    "TargetDatabase",
    "TargetFramework",
    "TargetLanguage",
    "TargetLibraries",
    "TargetPlatform",
    "TargetTools",
    "WorkedAIDeveloper",
    "WorkedAISearch",
    "z_BenefitsAI",
    "z_FavorableAI",
    "z_TrustAI",
]
df3 = df2.drop(columns_to_drop, axis="columns").copy()
df3.isna().sum()

Age                      0
Country                  0
CurrentJob             115
EdLevel                  0
Employment              12
Industry                 0
MainBranch               0
OSPersonal             485
OSProffesional        2989
RemoteWork              79
WorkedCollabTools      584
WorkedDatabase        6255
WorkedFramework      10154
WorkedLanguage         136
WorkedLibraries      16250
WorkedPlatform       10644
WorkedTools           3964
WorkingYears         15381
YearlySalary             0
YearsCode               69
YearsCodePro           194
z_UsingAI                0
dtype: int64

In [8]:
df3["MainBranch"].value_counts()
df3["Developer"] = df3["MainBranch"].apply(
    lambda x: 1 if x == "I am a developer by profession" else 0
)

In [9]:
df3[["WorkingYears", "YearsCodePro"]].corr()

WorkingYears  YearsCodePro
WorkingYears          1.00          0.93
YearsCodePro          0.93          1.00

Almost all the people are employed. I've already dealed with 'MainBranch' and 'WorkingYears' have too many missing values, and it is highly correlated with YearsCodePro, so I will drop it.

In [10]:
df3 = df3.drop(["Employment", "WorkingYears", "MainBranch"], axis="columns").copy()


In [11]:
df3["WorkedPlatform"].value_counts()
df3["WorkedFramework"].value_counts()
df3["WorkedLibraries"].value_counts()

WorkedLibraries
.NET (5+) ;.NET Framework (1.0 - 4.8)                                                                                                    2614
.NET (5+)                                                                                                                                2226
Spring Framework                                                                                                                         1471
React Native                                                                                                                              999
RabbitMQ                                                                                                                                  936
                                                                                                                                         ... 
Flutter;Keras;TensorFlow                                                                                                            

In [12]:
df4 = df3.drop(
    ["WorkedPlatform", "WorkedFramework", "WorkedLibraries"], axis="columns"
).copy()

df4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48019 entries, 1 to 87968
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                48019 non-null  object 
 1   Country            48019 non-null  object 
 2   CurrentJob         47904 non-null  object 
 3   EdLevel            48019 non-null  object 
 4   Industry           48019 non-null  object 
 5   OSPersonal         47534 non-null  object 
 6   OSProffesional     45030 non-null  object 
 7   RemoteWork         47940 non-null  object 
 8   WorkedCollabTools  47435 non-null  object 
 9   WorkedDatabase     41764 non-null  object 
 10  WorkedLanguage     47883 non-null  object 
 11  WorkedTools        44055 non-null  object 
 12  YearlySalary       48019 non-null  float64
 13  YearsCode          47950 non-null  float64
 14  YearsCodePro       47825 non-null  float64
 15  z_UsingAI          48019 non-null  object 
 16  Developer          48019 no

To predict a worker's salary, these seems to be the most important features.

In [13]:
df4["z_UsingAI"].value_counts()
df4["z_UsingAI"] = df4["z_UsingAI"].apply(lambda x: 1 if x == "Yes" else 0)

len(df4["EdLevel"].unique())
df4["EdLevel"].value_counts()
df4.tail()
df4.reset_index(drop=True, inplace=True)

In [14]:
df4["OSProffesional"].fillna(df4["OSPersonal"], inplace=True)
df4.drop("OSPersonal", axis=1, inplace=True)

df["WorkedCollabTools"].value_counts()
# This column is too messy, the most is VSCode, I don't think it will be useful

df4.drop("WorkedCollabTools", axis=1, inplace=True)

df4["WorkedDatabase"].value_counts().nlargest(10)

df5 = df4.copy()

I have some options: Drop the 6255 nan values or fill them with the most common value. But, I don't think in real life, the database and the tools you work it's going to be a good predictor. Instead of the language you work with. In this case, SQL could be a really good factor

In [15]:
df5.drop(["WorkedDatabase", "WorkedTools"], axis=1, inplace=True)

df5.isna().sum()

Age                 0
Country             0
CurrentJob        115
EdLevel             0
Industry            0
OSProffesional    193
RemoteWork         79
WorkedLanguage    136
YearlySalary        0
YearsCode          69
YearsCodePro      194
z_UsingAI           0
Developer           0
dtype: int64

In [16]:
df5.dropna(inplace=True)
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 47424 entries, 0 to 48018
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             47424 non-null  object 
 1   Country         47424 non-null  object 
 2   CurrentJob      47424 non-null  object 
 3   EdLevel         47424 non-null  object 
 4   Industry        47424 non-null  object 
 5   OSProffesional  47424 non-null  object 
 6   RemoteWork      47424 non-null  object 
 7   WorkedLanguage  47424 non-null  object 
 8   YearlySalary    47424 non-null  float64
 9   YearsCode       47424 non-null  float64
 10  YearsCodePro    47424 non-null  float64
 11  z_UsingAI       47424 non-null  int64  
 12  Developer       47424 non-null  int64  
dtypes: float64(3), int64(2), object(8)
memory usage: 5.1+ MB


In [18]:
# Remove YearlySalary outliers by the IQR method
df5.describe()
Q1 = df5["YearlySalary"].quantile(0.25)
Q3 = df5["YearlySalary"].quantile(0.75)
IQR = Q3 - Q1
df5 = df5[~((df5["YearlySalary"] < (Q1 - 1.5 * IQR))
        | (df5["YearlySalary"] > (Q3 + 1.5 * IQR))
    )]

In [19]:
df5.reset_index(drop=True, inplace=True)
df5

Age                                            Country  \
0      25-34 years old                           United States of America   
1      25-34 years old                                        Philippines   
2      35-44 years old  United Kingdom of Great Britain and Northern I...   
3      35-44 years old                           United States of America   
4      25-34 years old                           United States of America   
...                ...                                                ...   
44659  35-44 years old                                             Brazil   
44660  18-24 years old                                            Armenia   
44661  18-24 years old                                              India   
44662  25-34 years old                                             France   
44663  25-34 years old                                             Greece   

                    CurrentJob  \
0         Developer, front-end   
1        Developer, full-stack   
2          Developer, back-end   
3        Developer, full-stack   
4        Developer, full-stack   
...                        ...   
44659      Developer, back-end   
44660    Developer, full-stack   
44661        Developer, mobile   
44662  Other (please specify):   
44663    Developer, full-stack   

                                                 EdLevel            Industry  \
0           Bachelor’s degree (B.A., B.S., B.Eng., etc.)       Not_Responded   
1           Bachelor’s degree (B.A., B.S., B.Eng., etc.)               Other   
2      Some college/university study without earning ...               Other   
3      Some college/university study without earning ...          Technology   
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)  Financial Services   
...                                                  ...                 ...   
44659    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)          Technology   
44660  Some college/university study without earning ...  Financial Services   
44661       Bachelor’s degree (B.A., B.S., B.Eng., etc.)              Retail   
44662    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)               Other   
44663    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)          Technology   

                OSProffesional                            RemoteWork  \
0             iOS;iPadOS;MacOS  Hybrid (some remote, some in-person)   
1      Other (Please Specify):                                Remote   
2                        MacOS                                Remote   
3                        MacOS                                Remote   
4         MacOS;Ubuntu;Windows                                Remote   
...                        ...                                   ...   
44659                  Windows                                Remote   
44660          Red Hat;Windows                             In-person   
44661            Android;MacOS                                Remote   
44662                  Windows  Hybrid (some remote, some in-person)   
44663                  Windows                             In-person   

                                          WorkedLanguage  YearlySalary  \
0      Bash/Shell (all shells);HTML/CSS;JavaScript;PH...    156,000.00   
1                         HTML/CSS;JavaScript;TypeScript     23,456.00   
2      Bash/Shell (all shells);HTML/CSS;JavaScript;Ru...     96,828.00   
3      Ada;Clojure;Elixir;Go;HTML/CSS;Java;JavaScript...    135,000.00   
4      Go;HTML/CSS;JavaScript;Python;Rust;SQL;TypeScript     80,000.00   
...                                                  ...           ...   
44659                                              C;C++     50,719.00   
44660  Assembly;Bash/Shell (all shells);C;C#;C++;Java...     16,917.00   
44661                Dart;HTML/CSS;JavaScript;TypeScript     15,752.00   
44662           C#;F#;HTML/CSS;JavaScript;SQL;TypeScript     64,254.00   
44663                         Java;JavaScript

In [20]:
pd.to_pickle(df5, "../../data/interim/YearlySalary_model_p1.pkl")